compiler/src/iree/compiler/Dialect/LinalgExt/Utils/Utils.cpp

```diff
@@ -596,4 +598,180 @@ getIGEMMGenericConvDetails(linalg::LinalgOp linalgOp) {
+bool isaHorizontallyFusedContraction(linalg::LinalgOp linalgOp) {
+  if (linalgOp->getNumResults() == 1) {
+    return false;
+  }
+  // Check that the number of `ins` is one more than the number of results.
+  if (linalgOp.getNumDpsInputs() != linalgOp->getNumResults() + 1) {
+    return false;
+  }
+  SmallVector<AffineMap> indexingMaps = linalgOp.getIndexingMapsArray();
+  if (!llvm::all_of(indexingMaps, [](AffineMap m) {
+        return m.isProjectedPermutation() && !m.isPermutation();
+      })) {
+    return false;
+  }
+
+  llvm::SetVector<BlockArgument> rhsArgs;
+  llvm::SetVector<BlockArgument> outArgs;
+  for (auto [index, yieldedVal] :
+       llvm::enumerate(linalgOp.getBlock()->getTerminator()->getOperands())) {
+    std::optional<ContractionOpSequenceArgs> args =
+        isContractionOpSequence<arith::MulFOp, arith::AddFOp, arith::MulIOp,
+                                arith::AddIOp, complex::MulOp, complex::AddOp,
+                                arith::AndIOp, arith::OrIOp>(yieldedVal);
+    if (!args) {
+      return false;
+    }
+    BlockArgument lhs = args->operands.first;
+    BlockArgument rhs = args->operands.second;
+
+    // One of the block arguments must be argument 0, corresponding to the LHS.
+    if (lhs.getArgNumber() != 0) {
+      if (rhs.getArgNumber() != 0) {
+        return false;
+      }
+      std::swap(lhs, rhs);
+    }
+    assert(rhs.getArgNumber() != 0 && "cannot have rhs be arg number 0");
+    if (rhs.getArgNumber() != index + 1) {
+      return false;
+    }
+    BlockArgument accumulator = args->accumulator;
+    if (accumulator.getArgNumber() != index + linalgOp.getNumDpsInputs()) {
+      return false;
+    }
+  }
+
+  // Check that they have valid m, n and k dims.
+  ArrayRef<AffineMap> indexingMapsRef(indexingMaps);
+  AffineMap lhsIndexingMap = indexingMaps.front();
+
+  auto getResultDims = [](AffineMap m) {
+    auto r = llvm::map_range(m.getResults(), [](AffineExpr e) {
+      return cast<AffineDimExpr>(e).getPosition();
+    });
+    return llvm::SmallDenseSet<unsigned>(r.begin(), r.end());
+  };
+  llvm::SmallDenseSet<unsigned> lhsDims = getResultDims(lhsIndexingMap);
+
+  for (auto [rhsIndexingMap, outputIndexingMap] :
+       llvm::zip_equal(indexingMapsRef.slice(1, linalgOp.getNumDpsInputs() - 1),
+                       indexingMapsRef.take_back(linalgOp.getNumDpsInits()))) {
+    llvm::SmallDenseSet<unsigned> rhsDims = getResultDims(rhsIndexingMap);
+    llvm::SmallDenseSet<unsigned> outsDims = getResultDims(outputIndexingMap);
+    llvm::SmallDenseSet<unsigned> mDims = lhsDims;
+    llvm::set_intersect(mDims, outsDims);
+    if (mDims.empty()) {
+      return false;
+    }
+    llvm::SmallDenseSet<unsigned> nDims = rhsDims;
+    llvm::set_intersect(nDims, outsDims);
+    if (nDims.empty()) {
+      return false;
+    }
+    llvm::SmallDenseSet<unsigned> kDims = lhsDims;
+    llvm::set_intersect(kDims, rhsDims);
+    if (kDims.empty()) {
```

Comment:
If I'm reading the logic here properly, there is no requirement that all of the rhs indexing maps are equal, but the kernel config logic is only using the first pair of indexing maps. Can we strengthen/simplify this requirement to equality for all rhs and output maps?

Reply:
Actually the requirement is to allow rhs and outputs to have different indexing maps (only the LHS indexing map is the same for all the fused contractions). I think it is better to have an operation for horizontal fusion that will make this just be in the verifier.